pkg/etcdcli/cached_members.go

```diff
@@ -0,0 +1,123 @@
+package etcdcli
+
+import (
+	"context"
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	"sync"
+	"time"
+)
+
+const refreshPeriod = time.Minute
+
+type CachedMembers struct {
+	client  EtcdClient
+	lock    sync.RWMutex
+	members memberHealth
+
+	lastRefresh time.Time
+}
+
+func NewMemberCache(client EtcdClient) AllMemberLister {
+	return &CachedMembers{
+		client:      client,
+		lock:        sync.RWMutex{},
+		lastRefresh: time.UnixMilli(0),
+	}
+}
+
+func (c *CachedMembers) shouldRefresh() bool {
+	return c.lastRefresh.Add(refreshPeriod).Before(time.Now())
```

Comment:
before or after? Sorry I think this was my mistake. Can you add some unit tests for this refresh if you have the time please?

Reply:
i think before is correct, so the `lastRefresh` + `minutes` is before `now` 

i can add the test, but it is already tested extensively in `memberhealth_test` 

is there any other test cases i can add ?

Reply:
:+1: I think this whole struct needs _some_ unit testing, even if it's already covered - get creative!

Reply:
+1 :) 👍🏽